generic_server.cc

```diff
@@ -14,13 +14,103 @@
 #include <seastar/coroutine/parallel_for_each.hh>
 #include <seastar/core/reactor.hh>
 #include <seastar/core/smp.hh>
+#include <utility>
 
 #include "db/config.hh"
+#include "seastar/core/semaphore.hh"
 
 namespace generic_server {
 
-connection::connection(server& server, connected_socket&& fd, lw_shared_ptr<named_semaphore> sem)
-    : _conns_cpu_concurrency_semaphore(sem)
+class counted_data_source_impl : public data_source_impl {
+    data_source _ds;
+    connection::cpu_concurrency_t& _cpu_concurrency;
+
+    future<temporary_buffer<char>> invoke_with_counting(
+            std::function<future<temporary_buffer<char>>()> fun) {
+        if (_cpu_concurrency.stopped) {
+            return fun();
+        }
```

Comment:
While it's good you reduce the overhead after auth, just going through another layer is some overhead. Please measure it.

My preferred alternative is to implement nested scheduling groups, run auth inside a nested group, and implement a notifier for when a group runs out of tasks. It's quite a project though.

Reply:
There is also input_stream::detach() which could be used to unpeel the wrapper.

Reply:
```
   if (_buf) {
        throw std::logic_error("detach() called on a used input_stream");
    }
```
as I understand detach can't be used because it doesn't preserve the buffer

I will measure overhead

Reply:
I believe you're correct, it can only be used before starting consumption.

Ideas:
 - variation of detach that also returns the unconsumed buffer, and a way to create an input_stream from this buffer and the data source
 - data_source::unget() (but burdens everyone with the extra semantics)